### PR TITLE
[Messenger] Uses a messenger serializer, not an individual encoder/decoder

### PR DESCRIPTION
**Pull request by sroze**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | yes
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | ø
| License       | MIT
| Doc PR        | Will come

Makes the component-based even simpler.

**Before**
```php
        $encoderDecoder = Serializer::create();
        $middleware = [new SendMessageMiddleware(new SenderLocator([
            Message::class => new AmqpTransport($encoderDecoder, $encoderDecoder, $connection),
        ]))];
```

**After**
```php
       $middleware = [new SendMessageMiddleware(new SenderLocator([
            Message::class => new AmqpTransport(Serializer::create(), $connection),
        ]))];
```

### CODE REVIEW

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/messenger.xml [hunk 1]
```diff
@@ -62,7 +62,6 @@
 
         <service id="messenger.transport.amqp.factory" class="Symfony\Component\Messenger\Transport\AmqpExt\AmqpTransportFactory">
             <argument type="service" id="messenger.transport.encoder" />
```
Comment (sroze) [inline src/Symfony/Bundle/FrameworkBundle/Resources/config/messenger.xml]: This probably need to be changed to be `messenger.transport.serializer`.

File: src/Symfony/Component/Messenger/Transport/Serialization/SerializerInterface.php [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Messenger\Transport\Serialization;
+
+/**
+ * @author Samuel Roze <samuel.roze@gmail.com>
+ */
+interface SerializerInterface extends DecoderInterface, EncoderInterface
```
Comment (fabpot) [inline src/Symfony/Component/Messenger/Transport/Serialization/SerializerInterface.php]: Not sure we need to keep the decoder and encoder interface.

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/messenger.xml (hunk 1 above)
Comment (fabpot) [reply src/Symfony/Bundle/FrameworkBundle/Resources/config/messenger.xml]: indeed

File: src/Symfony/Component/Messenger/Transport/Serialization/SerializerInterface.php (hunk 2 above)
Comment (sroze) [reply src/Symfony/Component/Messenger/Transport/Serialization/SerializerInterface.php]: I've removed them 👍

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/messenger.xml (hunk 1 above)
Comment (sroze) [reply src/Symfony/Bundle/FrameworkBundle/Resources/config/messenger.xml]: Updated.

File: src/Symfony/Component/Messenger/Transport/Serialization/SerializerInterface.php [hunk 3]
```diff
@@ -27,8 +27,16 @@
...
      * - `body` (string) - the message body
      * - `headers` (string<string>) - a key/value pair of headers
-     *
-     * @return Envelope
      */
     public function decode(array $encodedEnvelope): Envelope;
+
+    /**
+     * Encodes an envelope content (message & items) to a common format understandable by transports.
+     * The encoded array should only contain scalar and arrays.
```
Comment (chalasr) [inline src/Symfony/Component/Messenger/Transport/Serialization/SerializerInterface.php]: `scalars and arrays`

**Review by chalasr — APPROVED** [review]
with minor comment

**Review by fabpot — APPROVED** [review]
@sroze Can you add an entry in the component CHANGELOG? And perhaps a note in UPGRADE file?

**Comment by sroze** [comment]
I've added changes on both CHANGELOG (Messenger & FWB) + the UPGRADE-4.2 👍

**Review by chalasr — APPROVED** [review]

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @sroze.

**Pull request merged**